insights/parsers/audit_log.py

```diff
@@ -0,0 +1,164 @@
+        type=LOGIN msg=audit(1506047401.407:53591): pid=482 uid=0 subj=system_u:system_r:crond_t:s0-s0:c0.c1023 old-auid=4294967295 auid=993 old-ses=4294967295 ses=7389 res=1
+        type=AVC msg=audit(1506487181.009:32794): avc:  denied  { create } for  pid=27960 comm="mongod" scontext=system_u:system_r:mongod_t:s0 tcontext=system_u:system_r:mongod_t:s0 tclass=unix_dgram_socket
+
+    Examples:
+        >>> log = shared[AuditLog]
+        >>> log.get('type=AVC')
+        [{
+            'timestamp': '1506487181.009',
+            'unparsed': 'avc:  denied  { create } for',
+            'msg_ID': '32794',
+            'pid': '27960',
+            'raw_message': 'type=AVC msg=audit(1506487181.009:32794): avc:  denied  { create } for  pid=27960 comm="mongod" scontext=system_u:system_r:mongod_t:s0 tcontext=system_u:system_r:mongod_t:s0 tclass=unix_dgram_socket',
+            'comm': 'mongod',
+            'scontext': 'system_u:system_r:mongod_t:s0',
+            'tclass': 'unix_dgram_socket',
+            'type': 'AVC',
+            'tcontext': 'system_u:system_r:mongod_t:s0'
+        }]
+        >>> assert len(list(log.get_after(timestamp=date.fromtimestamp(1506047401.407)))) == 3
+    """
+    def _parse_line(self, line):
+        """
+        Parse a log line into a info dictionary.
+
+        Parsing logic:
+
+            * First, use `shlex.split` to split by empty.
+            * Next, assert the first two key-value pair is 'type' and 'msg'.
+            * Next, parse the remained string reversly to get key-value pair data as more as possible.
+            * The left unparsed string will be stored at "unparsed".
+            * At last, wrap all key-value pairs into a dict called info.
+
+        Parameters:
+            line(string): Log line to parse
+
+        Yields:
+            Dictionary corresponding to the data in line split into
+            functional elements:
+
+            * **raw_message** - the original unparsed line
+            * **type** - the log type
+            * **timestamp** - the timestamp of the log line (as a string)
+            * **msg_ID** - the serial number behind the timestamp
+            * **unparsed** - the unparsed part in log, may or maynot exist
+
+            If the line is not regular, for some reason, then as many fields as
+            possible are pulled from the line.
+        """
+        info = {}
+        linesp = shlex.split(line)
+
+        if (len(linesp) < 2 or
+                not (linesp[0] and linesp[0].startswith('type=')) or
+                not (linesp[1] and linesp[1].startswith('msg=audit('))):
+            return
+
+        info['type'] = linesp[0].lstrip('type=')
+
+        timestamp_id = linesp[1].lstrip('msg=audit(').rstrip('):')
+        timestamp_id_sp = timestamp_id.split(':')
+        if len(timestamp_id_sp) != 2:
+            return
```

Comment:
Better return the `raw_message`, no matter it can be parsed or not.

Reply:
Add a flag named `is_valid` and return `{'raw_message': line, 'is_valid': False} ` if unparseable.
```